servant-server/src/Servant/Server/Internal/Authentication.hs

```diff
@@ -0,0 +1,131 @@
+{-# LANGUAGE CPP                     #-}
+{-# LANGUAGE DataKinds               #-}
+{-# LANGUAGE FlexibleInstances       #-}
+{-# LANGUAGE FunctionalDependencies  #-}
+{-# LANGUAGE MultiParamTypeClasses   #-}
+{-# LANGUAGE OverloadedStrings       #-}
+{-# LANGUAGE ScopedTypeVariables     #-}
+{-# LANGUAGE TypeFamilies            #-}
+
+module Servant.Server.Internal.Authentication
+( AuthData (..)
+, authProtect
+, basicAuthLax
+, basicAuthStrict
+, jwtAuthStrict
+, SimpleAuthProtected
+        ) where
+
+#if !MIN_VERSION_base(4,8,0)
+import           Control.Applicative        ((<$>))
+#endif
+import           Control.Monad              (guard)
+import qualified Data.ByteString            as B
+import           Data.ByteString.Base64     (decodeLenient)
+import           Data.Monoid                ((<>))
+import           Data.Proxy                 (Proxy (Proxy))
+import           Data.String                (fromString)
+import           Data.Word8                 (isSpace, toLower, _colon)
+import           GHC.TypeLits               (KnownSymbol, symbolVal)
+import           Data.Text.Encoding         (decodeUtf8)
+import           Data.Text                  (splitOn)
+import           Network.Wai                (Request, requestHeaders)
+import Servant.Server.Internal.ServantErr (err401, ServantErr(errHeaders))
+import           Servant.API.Authentication (AuthPolicy (Strict, Lax),
+                                             AuthProtected(..),
+                                             BasicAuth (BasicAuth),
+                                             JWTAuth(..),
+                                             OnMissing (..),
+                                             OnUnauthenticated (..))
+import            Web.JWT                    (decodeAndVerifySignature, JWT, VerifiedJWT, Secret)
+import qualified  Web.JWT as JWT             (decode)
+
+-- | Class to represent the ability to extract authentication-related
+-- data from a 'Request' object.
+class AuthData a e | a -> e where
+    authData :: Request -> Either e a
+
+-- | combinator to create authentication protected servers.
+authProtect :: OnMissing IO ServantErr missingPolicy missingError
```

Comment:
yeah, I use it. I find it more idiomatic when defining the api to a server to use a function rather than a constructor, but ymmv.